### ISSUE TITLE
FR: Annotate open file with PR # which would change that file if merged

### ISSUE THREAD

**Issue by katrinleinweber** [KnisterPeter/vscode-github#480]
Hello & thanks for publishing this extension :-)

`git blame` lets us find out which commit(ter) last touched a given line. Thus it's basically like looking into the past. But what about the future of a file?

Would it be feasible to add a kinda forward-looking, blame-like mode which would check open PRs whether they are modifying a file and if yes, annotate the to-be-changed lines?

My goal is to be alerted before starting duplicate work, or risking merge conflicts. Example use case: proof-reader finds some typo, but sees that there is already `PR #...` open against that file/line, thus doesn't need to send another PR that fixes the typo.

I was directed here by https://github.com/eamodio/vscode-gitlens/issues/251, and having looked through the existing issues I only find any about the classic blame view (like #214). I looked through [Microsoft's `vscode-pull-request-github` issues](https://github.com/Microsoft/vscode-pull-request-github/issues) but didn't find something that sounded similar to this FR. If something like this exists already elsewhere, I'd be happy about a pointer.

Cheers :-)

**Comment by KnisterPeter** [comment]
Hi @katrinleinweber, that sounds like an interesting and helpful feature. I don't have much time right now to implement this but I would greatly appreciate any start and add it to this extension.

**Comment by katrinleinweber** [comment]
How much Java/TypeScript experience do you think is needed to implement this? I'm starting with almost none, so I can't promise anything here :-(

However, I found @dzhavat's [Potential changes for GitHub](https://github.com/dzhavat/potential-changes-for-github) which does what I described above on the file level, which is a useful start for me :-)

Maybe his JS code could be the start of a per-line-solution to this issue?

**Comment by katrinleinweber** [comment]
Ah! Great minds think alike: https://github.com/sindresorhus/refined-github/pull/1691 :-)

**Comment by KnisterPeter** [comment]
@katrinleinweber  I think this needs a bit of work, since you are required to fetch all open PR, test if the currently open file is affected, decode the diffs in the PR and then annotate the sources of the open file.
Its not easy and may be a lot of work, but I don't want to stop you or frustrate you. Its really a very nice idea.
I will help you with any questions which may arise.

**Comment by dzhavat** [comment]
> However, I found @dzhavat's Potential changes for GitHub which does what I described above on the file level, which is a useful start for me :-)

@katrinleinweber Thanks for mentioning my extension. It's a very humble try to create something useful. What you describe in your initial comment is exactly the reason I created this - to avoid duplicate work.

You're welcome to use all of the code from my extension. I'm actually in the process of implementing personal tokens which will enable me to use GraphQL, which will bring down the number of requests significantly.

If I'm not mistaken, this extension already has an implementation for personal so fetching all open PRs shouldn't be that big of a problem (unless there are quite many of them (like a few hundred)). Testing whether the current file is affected by a PR is not big of a problem either (at least from my experience). I'm not really sure about the rest of the work though.

I think this kind of extension/feature is really needed and it can turn out to be quite valuable. I, too, am ready to help with any questions you might have.

**Comment by katrinleinweber** [comment]
Just for completeness' sake: Refined GitHub has implemented the [list-prs-for-file](https://github.com/sindresorhus/refined-github/blob/600ecbfd830702181cd800516630680a8a0507fb/source/features/list-prs-for-file.tsx) feature, which notifies users when editing a file that other PRs are already touching. Slightly different use-case, but same logic underneath.
